url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/math/MathExpression.java

```diff
@@ -0,0 +1,615 @@
+		for (final HashSet<MathsItem> m : items) {
+			final HashSet<MathsItem> clone = new HashSet<MathsItem>();
+			
+			for (final MathsItem it : m)
+				clone.add(it.clone());
+			
+			dest.add(clone);
+		}
+	}
+	
+	@Override
+	public String toString() {
+		
+		final StringBuilder ans = new StringBuilder();
+		
+		ans.append("(");
+		if (DEBUG) ans.append(" [Exp]");
+		
+		boolean first = true;
+		
+		for (HashSet<MathsItem> multiples : items) {
+			
+			if (multiples.isEmpty())
+				continue;
+			
+			boolean first2 = true;
+			
+			for (MathsItem it : multiples) {
+				
+				final boolean itnegative = it.hasNegativeSign();
+				
+				if (!first2)
+					ans.append("*");
+				else if (DEBUG || (!itnegative && !first))
+					ans.append("+");
+				
+				if (!first2 && itnegative && multiples.size() > 1) ans.append("(");
+				ans.append(it);
+				if (!first2 && itnegative && multiples.size() > 1) ans.append(")");
+	
+				first2 = false;
+			}
+			
+			first = false;
+				
+		}
+		
+		ans.append(")");
+		
+		return ans.toString();
+	}
+	
+	public static MathExpression fromString(final String input) throws Exception {
+		
+		final ArrayList<Tuple<char[], String>> els = Tools.splitByTopLevel(input, new char[]{'+', '-'}, true);
+	
+		if (els == null) {
+			// dirty hack to fix inputs like "2*3"
+			
+			final ArrayList<Tuple<char[], String>> temp = Tools.splitByTopLevel(input, new char[]{'*'}, false);
+			
+			if (temp != null && !temp.isEmpty())
+				return fromString("0+"+input);
+			else
+				return null;
+		}
+		
+		//if (els.size() < 2) return null;
+		
+		final MathExpression ans = new MathExpression();
+
+		for (final Tuple<char[], String> ms : els)
+			ans.items.add(parsemultiples(ms));
+
+		ans.simplify();
+		
+		return ans;
+	}
+	
+	@Override
+	public void complexconjugate() {
+		for (final HashSet<MathsItem> its : items)
+			for (final MathsItem it : its)
+					it.complexconjugate();
+	}
+	
+	private static HashSet<MathsItem> parsemultiples(final Tuple<char[], String> ms) throws Exception {
+		final ArrayList<Tuple<char[], String>> multiples = Tools.splitByTopLevel(ms.y, new char[]{'*'}, false);
+		final HashSet<MathsItem> forstore = new HashSet<MathsItem>();
+		
+		int sign = 1;
+		
+		for (int i = 0; i < ms.x.length; i++)
+			if (ms.x[i] == '-')
+				sign *= -1;
+		
+		if (multiples != null) {
+
+			boolean first = true;
+			
+			for (final Tuple<char[], String> it : multiples) {
+				final MathsItem mit = MathsParser.parse(it.y);
+				if (first && sign < 1) mit.negate();
+				forstore.add(mit);
+				first = false;
+			}
+		} else {
+			final MathsItem mit = MathsParser.parse(ms.y);
+			if (sign < 1) mit.negate();
+			forstore.add(mit);
+		}
+		
+		return forstore;
+	}
+	
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof MathsItem)
+			return getValue(null).equals(((MathsItem) obj).getValue(null));
+		
+		return false;
+	}
+
+	@Override
+	public boolean divide(MathsItem m) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.